api/src/main/java/com/github/eyefloaters/console/api/model/Node.java

```diff
@@ -4,10 +4,47 @@
 import com.fasterxml.jackson.annotation.JsonInclude.Include;
 
 @JsonInclude(value = Include.NON_NULL)
-public record Node(int id, String host, int port, String rack) {
+public class Node {
+
+    private final int id;
+    private final String host;
+    private final int port;
+    private final String rack;
+    private Either<ReplicaInfo, Error> log;
```

Comment:
A node hosts many replicas, so why is there just a single one here? TBH it's not clear to me why it's better to model the relationship between nodes and replicas on the Node, rather than on the Partition.

Reply:
This goes back to my earlier comment about the `Node` model. Here, a `Node` represents the placement of a partition replica. Thus, the `log` is the storage used by that partition's replica on the node. As you're pointing out, this is not clear with the way it's modeled currently - or at least the way the models are named.

Reply:
I did wonder about how many node instances there were floating about for a given node id! If you wanted to stick with this way of doing it then we should change the name `Node`→`ReplicaNode` or something.

Reply:
Suppose we do something like this:
```java
// replaces `Node` in the context of describing a replica
class Replica {
  int nodeId;
  String nodeRack;
  boolean leader;
  boolean inSync;
  // localStorage? Depends on if/how we can get tiered storage info
  Either<ReplicaStorage, Error> storage;
}

// replaces `ReplicaInfo` (too generic?)
record ReplicaStorage(
  long localSize,
  long offsetLag,
  boolean future
  ) {
}
```

The partition model could then simply have a `List` of `Replica`s, and get rid of `leader` and `isr` as separate fields.

Reply:
I would keep `int leaderId` in the partition model, then you could use `List<Replica> replicas`, in the same order as returned from the Admin client (thus preserving semantic of the first replica being the preferred leader). Of course you can encapsulate the lookup of the leader (on the partition model) via its id in a method so the caller doesn't have to know there isn't a `Replica leader` field.

I'd also wonder about inlining the fields of `ReplicaStorage` into `Replica` with a separate `Error storageError`. Doing it as you suggest means the structure of the objects reflects how the information was retrieved (i.e. it's only because we use a separate Admin method for getting the storage into that we have a separate class, and a indirection through an `Either`), which feels pretty invasive.  It's one thing to do that in your Java code, but when it leaks out into the JSON structure exposed by the API it starts to feel like we're exposing internal details of the implementation. At that point it might be cleaner if all the fields were nullable and each class (`Replica` etc) had a single `Error` which was non-null  whenever there was any problem. But that comes with it's own downsides for writing the code in a functional way. WDYT?

Reply:
I do think there's a benefit to placing things in the `Either`, especially when it comes to a set of fields that may be inaccessible due to permissions or some combination of ACLs preventing the retrieval. Granted it does expose some info about the backend, but ultimately it does come down to the API operations that Kafka exposes, regardless of any server's implementation. 

If we later add some additional `remoteStorage` for Tiered Storage information, it compartmentalizes that storage info or the reason it is not present neatly beneath a single field in the model.

Reply:
As mentioned in the other thread, a new object `PartitionReplica` was added to carry the replica `nodeId` and `nodeRack` information, plus a boolean derived from the `isr` array. This embeds an `Either<ReplicaLocalStorage, Error>` for node-local storage of the log. I'll leave this thread open in case the discussion around how/whether to use `Either` like this should continue.